File: plus-one.java
public class Solution {
    public int[] plusOne(int[] digits) {
        int carry = 1;
        for (int i = digits.length - 1; i >= 0 && carry > 0; i--) {
            digits[i] += carry;
            carry = digits[i] / 10;
            digits[i] = digits[i] % 10;
        }
        if (carry > 0) {
            int[] ans = new int[digits.length + 1];
            for (int i = 0; i < digits.length; i++)
                ans[i + 1] = digits[i];
            ans[0] = carry;
            return ans;
        }
        return digits;
    }
}

File: plus-one.cc
class Solution {
public:
    vector<int> plusOne(vector<int> &digits) {
        reverse(digits.begin(), digits.end());
        int carry = 1;
        for (int i = 0; i < digits.size() && carry > 0; i++) {
            digits[i] += carry;
            if (digits[i] == 10) {
                digits[i] = 0;
                carry = 1;
            } else {
                carry = 0;
            }
        }
        if (carry == 1) digits.push_back(1);
        reverse(digits.begin(), digits.end());
        return digits;
    }
};
